File: src/Archive/ArchiveParser.cs
using System.Text.Json;
using System.Text.RegularExpressions;

namespace Pyxelze;

internal static class ArchiveParser
{
    private static readonly Regex LineSizeRegex = new(@"^(.+?)\s+\((\d+)\s+bytes\)$", RegexOptions.Compiled);

    public static List<VirtualFile> Parse(string data)
    {
        if (string.IsNullOrWhiteSpace(data)) return new();

        data = data.Trim();
        return IsJsonArray(data) ? ParseJson(data) : ParseText(data);
    }

    public static List<string> ParseFileNames(string data)
    {
        if (string.IsNullOrWhiteSpace(data)) return new();
        data = data.Trim();

        List<string> names;
        if (IsJsonArray(data))
            names = ParseJsonNames(data);
        else
            names = data.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
                .Select(line => LineSizeRegex.Match(line.Trim()))
                .Where(m => m.Success)
                .Select(m => m.Groups[1].Value.Trim())
                .ToList();

        return names;
    }

    private static bool IsJsonArray(string data) =>
        data.StartsWith("[") && data.EndsWith("]");

    private static List<string> ParseJsonNames(string data)
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return new();
            return doc.RootElement.EnumerateArray()
                .Where(e => e.ValueKind == JsonValueKind.Object && e.TryGetProperty("name", out _))
                .Select(e => e.GetProperty("name").GetString())
                .Where(n => !string.IsNullOrEmpty(n))
                .ToList()!;
        }
        catch { return new(); }
    }

    private static List<VirtualFile> ParseJson(string data)
    {
        var rawPaths = new List<(string name, long size)>();
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (doc.RootElement.ValueKind != JsonValueKind.Array) return new();

            foreach (var file in doc.RootElement.EnumerateArray())
                rawPaths.Add((file.GetProperty("name").GetString() ?? "", file.GetProperty("size").GetInt64()));
        }
        catch { return new(); }

        var result = new List<VirtualFile>();
        foreach (var (name, size) in rawPaths)
        {
            if (string.IsNullOrEmpty(name)) continue;

            result.Add(new VirtualFile
            {
                FullPath = name,
                OriginalPath = name,
                Name = Path.GetFileName(name),
                Size = size,
                IsFolder = false
            });

            AddParentDirectories(result, name);
        }
        return result;
    }

    private static List<VirtualFile> ParseText(string data)
    {
        var rawPaths = new List<(string path, long size)>();
        foreach (var line in data.Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries))
        {
            var trimmed = line.Trim();
            if (trimmed.StartsWith("Files in")) continue;

            var match = LineSizeRegex.Match(trimmed);
            if (!match.Success) continue;

            rawPaths.Add((match.Groups[1].Value.Trim().Replace("\\", "/"), long.Parse(match.Groups[2].Value)));
        }

        var result = new List<VirtualFile>();
        foreach (var (path, size) in rawPaths)
        {
            if (string.IsNullOrEmpty(path)) continue;

            result.Add(new VirtualFile
            {
                FullPath = path,
                OriginalPath = path,
                Name = Path.GetFileName(path),
                Size = size,
                IsFolder = false
            });

            AddParentDirectories(result, path);
        }
        return result;
    }

    private static void AddParentDirectories(List<VirtualFile> files, string filePath)
    {
        var dir = Path.GetDirectoryName(filePath)?.Replace("\\", "/");
        while (!string.IsNullOrEmpty(dir))
        {
            if (!files.Any(f => f.FullPath == dir))
            {
                files.Add(new VirtualFile
                {
                    FullPath = dir,
                    Name = Path.GetFileName(dir) ?? "",
                    IsFolder = true
                });
            }
            dir = Path.GetDirectoryName(dir)?.Replace("\\", "/");
        }
    }
}

File: src/archive/ArchiveParser.cpp
#include "ArchiveParser.h"
#include <QJsonDocument>
#include <QJsonArray>
#include <QJsonObject>
#include <QRegularExpression>
#include <QFileInfo>
#include <QSet>

static const QRegularExpression LineSizeRegex(QStringLiteral(R"(^(.+?)\s+\((\d+)\s+bytes\)$)"));

static bool isJsonArray(const QString& d) {
    return d.startsWith('[') && d.endsWith(']');
}

static void addParentDirs(QList<VirtualFile>& files, const QString& filePath, QSet<QString>& seen) {
    auto dir = QFileInfo(filePath).path().replace('\\', '/');
    while (!dir.isEmpty() && dir != QStringLiteral(".")) {
        if (seen.contains(dir)) break;
        seen.insert(dir);
        VirtualFile vf;
        vf.fullPath = dir;
        vf.name = QFileInfo(dir).fileName();
        vf.isFolder = true;
        files.append(vf);
        dir = QFileInfo(dir).path().replace('\\', '/');
    }
}

static QList<VirtualFile> parseJson(const QString& data) {
    QList<VirtualFile> result;
    QSet<QString> seenDirs;
    auto doc = QJsonDocument::fromJson(data.toUtf8());
    if (!doc.isArray()) return result;

    for (const auto& val : doc.array()) {
        auto obj = val.toObject();
        auto name = obj[QStringLiteral("name")].toString();
        auto size = obj[QStringLiteral("size")].toInteger();
        if (name.isEmpty()) continue;

        VirtualFile vf;
        vf.fullPath = name;
        vf.originalPath = name;
        vf.name = QFileInfo(name).fileName();
        vf.size = size;
        vf.isFolder = false;
        result.append(vf);
        addParentDirs(result, name, seenDirs);
    }
    return result;
}

static QList<VirtualFile> parseText(const QString& data) {
    QList<VirtualFile> result;
    QSet<QString> seenDirs;

    for (const auto& line : data.split('\n', Qt::SkipEmptyParts)) {
        auto trimmed = line.trimmed();
        if (trimmed.startsWith(QStringLiteral("Files in"))) continue;

        auto match = LineSizeRegex.match(trimmed);
        if (!match.hasMatch()) continue;

        auto path = match.captured(1).trimmed().replace('\\', '/');
        auto size = match.captured(2).toLongLong();

        VirtualFile vf;
        vf.fullPath = path;
        vf.originalPath = path;
        vf.name = QFileInfo(path).fileName();
        vf.size = size;
        vf.isFolder = false;
        result.append(vf);
        addParentDirs(result, path, seenDirs);
    }
    return result;
}

QList<VirtualFile> ArchiveParser::parse(const QString& data) {
    auto d = data.trimmed();
    if (d.isEmpty()) return {};
    return isJsonArray(d) ? parseJson(d) : parseText(d);
}

QStringList ArchiveParser::parseFileNames(const QString& data) {
    auto d = data.trimmed();
    if (d.isEmpty()) return {};

    if (isJsonArray(d)) {
        QStringList names;
        auto doc = QJsonDocument::fromJson(d.toUtf8());
        if (!doc.isArray()) return {};
        for (const auto& val : doc.array()) {
            auto n = val.toObject()[QStringLiteral("name")].toString();
            if (!n.isEmpty()) names.append(n);
        }
        return names;
    }

    QStringList names;
    for (const auto& line : d.split('\n', Qt::SkipEmptyParts)) {
        auto match = LineSizeRegex.match(line.trimmed());
        if (match.hasMatch())
            names.append(match.captured(1).trimmed());
    }
    return names;
}
